robot_3d/src/joint_states_subscriber_3d.cpp

```diff
@@ -0,0 +1,233 @@
+
+  KDL::ChainFkSolverPos_recursive fksolver = KDL::ChainFkSolverPos_recursive(chain);
+
+  // Creating a kdl array to store joint positions.
+  unsigned int nj = chain.getNrOfJoints();
+  KDL::JntArray jointpositions = KDL::JntArray(nj);
+  ROS_INFO("NrOfJoints =%d", nj);
+
+  // assign joint positions...
+  for (unsigned int i = 0; i < nj; i++)
+  {
+    jointpositions(i) = joint_positions[i];
+  }
+
+  // Create the frame that will contain the end effector pose.
+  KDL::Frame cartpos;
+
+  // Calculate forward kinematics for end effector.
+  bool kinematics_status;
+  kinematics_status = fksolver.JntToCart(jointpositions, cartpos);
+  if (kinematics_status >= 0)
+  {
+    std::cout << cartpos << std::endl;
+    ROS_INFO("%s \n", "Success, thanks KDL!");
+  }
+  else
+  {
+    ROS_INFO("%s \n", "Error: could not calculate forward kinematics :(");
+  }
+
+  // Creating and broadcasting tf transform for end effector.
+  tf::Transform transform;
+  tf::transformKDLToTF(cartpos, transform);
+
+  static tf::TransformBroadcaster br;
+  br.sendTransform(tf::StampedTransform(transform, ros::Time::now(), "base_link", "test_frame"));
+
+  // Outputting data to yaml file.
+  outputYaml(transform, joint_positions);
+}
+
+std::vector<double> JointMeasurementData::getJointOffsets(double joint_positions[])
+{
+  ros::NodeHandle n;
+
+  // Vector to store the offset joint angles for robot.
+  std::vector<double> offset_joints;
+
+  n.getParam("offsets", offset_joints);
+
+  // Adding offsets to joint angles.
+  for (unsigned i = 0; i < offset_joints.size(); i++)
+  {
+    offset_joints[i] += joint_positions[i];
+  }
+
+  return offset_joints;
+}
+
+void JointMeasurementData::outputYaml(tf::Transform kdl_transform, double joint_positions[])
+{
+  // Storing the position and orientation of end effector.
+  tf::Vector3 translation = kdl_transform.getOrigin();
+  tf::Quaternion orientation = kdl_transform.getRotation();
+
+  // Offset joint angles.
+  std::vector<double> offset_joint_angles = getJointOffsets(joint_positions);
+
+  // Outputting yaml stuff
+  std::stringstream stream;
+
+  // Parsing the data into yaml format.
+  YAML::Emitter d_output;
+  d_output << YAML::BeginSeq;
+  d_output << YAML::BeginMap << YAML::Key << "end_effector_translation" << YAML::Value << YAML::Flow << YAML::BeginSeq
+           << translation[0] << translation[1] << translation[2] << YAML::EndSeq;
+  d_output << YAML::Key << "end_effector_orientation" << YAML::Value << YAML::Flow << YAML::BeginSeq << orientation[0]
+           << orientation[1] << orientation[2] << orientation[3] << YAML::EndSeq;
+  d_output << YAML::Key << "offset_joint_angles" << YAML::Value << YAML::Flow << offset_joint_angles << YAML::EndSeq
+           << YAML::EndMap;
+  d_output << YAML::EndSeq;
+  // Storing the parsed data.
+  output_data_.append(d_output.c_str());
+  output_data_.append("\n");
+
+  // Outputting to file.
+  std::ofstream fout;
+  fout.open(output_filepath_);
```

Comment:
You reopen your file everytime you want to write and it looks like you never close. This is dangerous IO. I recommend that you open the file in the ctor and close in the dtor.